Lower the MSRV

The current version needs a minimum supported Rust version of 1.75 due to the `Links` trait returning an iterator.

As such a recent MSRV shouldn't be a blocker for anyone using this library, but it currently is, (see https://github.com/ipld/rust-ipld-core/pull/11 for more). We should find a way to lower the MSRV.

I currently see two ways (if anyone has others, please post here):

 1. Returning a boxed trait instead of an impl trait. For an implementation see #12.
    - Pro: Also the `Codec` trait would work on lower versions
    - Con:
      - It's a breaking change
      - It uses a dynamic trait 
 2. Introduce a `codec` feature flag, which is enabled but default. Users that don't need the `Codec` trait could just disable that feature and use a lower MSRV.
    - Pro: Non-breaking change (as long as you use the default features)
    - Con:
       - You cannot use the `Codec` trait in Rust < 1.75

> I'd prefer to keep it as-is or return an associated type (where implementers can choose to box).

@Stebalien I hope I understand it correctly that having an associated type would looke like this:

```rust
pub trait Links {
    type LinksError;
    type Iter;
    fn links(bytes: &[u8]) -> Result<Self::Iter, Self::LinksError>;
}
```

Returning a boxed iterator work, but `impl Iteractor` does not:

```console
error[E0658]: `impl Trait` in associated types is unstable
  --> src/codec.rs:37:17
   |
37 |     type Iter = impl Iterator<Item = Cid>;
   |                 ^^^^^^^^^^^^^^^^^^^^^^^^^
   |
   = note: see issue #63063 <https://github.com/rust-lang/rust/issues/63063> for more information

For more information about this error, try `rustc --explain E0658`.
```

So if you would prefer keeping it as-is, would adding the `codec` feature a possible solution you could at least leave with? The the `serde_ipld_dagcbor` would then also need such a feature. I really want to unblock users of this library and I can see the 1.75 might be too new (although indeed IPLD is bleeding edge).

--- a/Cargo.toml
+++ b/Cargo.toml
@@ -11,13 +11,15 @@ license = "MIT OR Apache-2.0"
 categories = ["data-structures", "encoding"]
 
 [features]
-default = ["std"]
+default = ["codec", "std"]
 # Makes the error implement `std::error::Error` and the `Codec` trait available.
 std = ["cid/std", "serde?/std", "serde_bytes?/std"]
 # Enables support for Serde serialization into/deserialization from the `Ipld` enum.
 serde = ["dep:serde", "dep:serde_bytes", "cid/serde"]
 # Enables support for property based testing.
 arb = ["dep:quickcheck", "cid/arb"]
+# Enables support for the Codec trait, needs at least Rust 1.75
+codec = []
 
 [dependencies]
 cid = { version = "0.11.1", default-features = false, features = ["alloc"] }

--- a/README.md
+++ b/README.md
@@ -85,7 +85,8 @@ fn main() {
 Feature flags
 -------------
 
- - `std`: Makes the error implement `std::error::Error` and the `Codec` trait available.
+ - `std` (enabled by default): Makes the error implement `std::error::Error` and the `Codec` trait available.
+ - `codec` (enabled by default): Provides the `Codec` trait, which enables encoding and decoding independent of the IPLD Codec. The minimum supported Rust version (MSRV) can significantly be reduced to 1.64 by disabling this feature.
  - `serde`: Enables support for Serde serialization into/deserialization from the `Ipld` enum.
  - `arb`: Enables support for property based testing.
 

--- a/src/lib.rs
+++ b/src/lib.rs
@@ -5,7 +5,7 @@
 
 extern crate alloc;
 
-#[cfg(feature = "std")]
+#[cfg(all(feature = "std", feature = "codec"))]
 pub mod codec;
 pub mod convert;
 pub mod ipld;

